lib/Backend/amd64/LowererMDArch.cpp

```diff
@@ -857,22 +881,85 @@ LowererMDArch::LowerCall(IR::Instr * callInstr, uint32 argCount)
     //
 
     AssertMsg(this->helperCallArgsCount >= 0, "Fatal. helper call arguments ought to be positive");
-    AssertMsg(this->helperCallArgsCount < 255, "Too many helper call arguments");
+    AssertMsg(this->helperCallArgsCount < MaxArgumentsToHelper && MaxArgumentsToHelper < 255, "Too many helper call arguments");
 
     uint16 argsLeft = static_cast<uint16>(this->helperCallArgsCount);
 
+    // Sys V x64 ABI assigns int and xmm arg registers separately.
+    // e.g. args:   int, double, int, double, int, double
+    //  Windows:    int0, xmm1, int2, xmm3, stack, stack
+    //  Sys V:      int0, xmm0, int1, xmm1, int2, xmm2
+#ifdef _WIN32
+#define _V_ARG_INDEX(index) index
+#else
+    uint16 _vindex[MaxArgumentsToHelper];
+    {
+        uint16 intIndex = 1, doubleIndex = 1, stackIndex = IntArgRegsCount + 1;
+        for (int i = 0; i < this->helperCallArgsCount; i++)
+        {
+            IR::Opnd * helperSrc = this->helperCallArgs[this->helperCallArgsCount - 1 - i];
+            IRType type = helperSrc->GetType();
+            if (IRType_IsFloat(type) || IRType_IsSimd128(type))
+            {
+                if (doubleIndex <= XmmArgRegsCount)
+                {
+                    _vindex[i] = doubleIndex++;
+                }
+                else
+                {
+                    _vindex[i] = stackIndex++;
+                }
+            }
+            else
+            {
+                if (intIndex <= IntArgRegsCount)
+                {
+                    _vindex[i] = intIndex++;
+                }
+                else
+                {
+                    _vindex[i] = stackIndex++;
+                }
+            }
+        }
+    }
+#define _V_ARG_INDEX(index) _vindex[(index) - 1]
+#endif
+
     while (argsLeft > 0)
     {
         IR::Opnd * helperSrc = this->helperCallArgs[this->helperCallArgsCount - argsLeft];
-        StackSym * helperSym = m_func->m_symTable->GetArgSlotSym(static_cast<uint16>(argsLeft));
-        helperSym->m_type = helperSrc->GetType();
+        uint16 index = _V_ARG_INDEX(argsLeft);
+        StackSym * helperSym = m_func->m_symTable->GetArgSlotSym(index);
+        helperSym->m_type = ExtendHelperArg(helperSrc->GetType());
         Lowerer::InsertMove(
-            this->GetArgSlotOpnd(argsLeft, helperSym),
+            this->GetArgSlotOpnd(index, helperSym, /*isHelper*/true),
             helperSrc,
             callInstr);
         --argsLeft;
     }
 
+#ifndef _WIN32
+    // Manually home args
+    if (argCount > 0)
+    {
+        static const RegNum s_argRegs[IntArgRegsCount] = {
+    #define REG_INT_ARG(Index, Name)  Reg ## Name,
+    #include "RegList.h"
+        };
+
+        const int callArgCount = static_cast<int>(argCount) + this->helperCallArgsCount;
```

Comment:
@rajatd I looked at this issue. Yes helperCallArgsCount can be non-zero. The reason is that many lowering code uses LowerCallArgs() to lower variadic JS runtime args, and LoadHelperArgument() to lower other "known args". So the full args length is actually their sum, while `argCount > 0` is an indicator that we need variadic args support so we need to manually home args.

Your concern of manual homing args here overwriting above helperCallArgs stack slots could potentially happen... e.g. `helperCallArgs == 7` so it has an arg at stack[0]. I can fix this by changing above `/*isHelper*/true` to `argCount == 0` so that their assumptions are consistent.